src/poetry/core/packages/specification.py

```diff
@@ -151,19 +151,24 @@ def is_same_package_as(self, other: PackageSpecification) -> bool:
 
         return self.is_same_source_as(other)
 
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, PackageSpecification):
+            return NotImplemented
+        return self.is_same_package_as(other)
+
     def __hash__(self) -> int:
-        if not self._source_type:
-            return hash(self._name)
-
-        return (
-            hash(self._name)
-            ^ hash(self._source_type)
-            ^ hash(self._source_url)
-            ^ hash(self._source_reference)
-            ^ hash(self._source_resolved_reference)
-            ^ hash(self._source_subdirectory)
-            ^ hash(self._features)
-        )
+        result = hash(self.complete_name)  # complete_name includes features
+
+        if self._source_type:
+            # don't include _source_reference and _source_resolved_reference in hash
```

Comment:
I do not think a spec with reference "master" and "v1.0.0" should be considered equal as they are inherently different. I wonder if resovled reference is what we do not need.

Reply:
Your example is not considered equal. 😉 See https://github.com/python-poetry/poetry-core/blob/6b45d6c3977d1963d7a247ebbc64d13e7fc2879e/src/poetry/core/packages/specification.py#L112-L146

However, this logic allows references to be different within certain limitations:

- If `source_reference` is not set, two specs with arbitrary `resolved_source_reference` can be equal. (Might be a theoretical use case.)
- If `source_reference` is set, it is allowed that one starts with the other if `resolved_source_reference` is equal.

See the test added in `test_specification.py` for example. (I just added a sentence for clarity in the comment and another test for the theoretical use case.)

Reply:
Wait a minute, this logic is partially redundant and can be further simplified.

Reply:
OK, nothing redundant in the logic. The test now checks each variant that is considered equal. Not sure if the logic for references in `is_same_source_as()` completely makes sense, but that may be deferred to another PR. I suppose for now, it may be "good enough" (since the focus of this PR is consistency between `__eq__` and `__hash__`).

If the logic is changed later, we can still consider to include additional attributes into the hash.

Reply:
I am thinking of the caches we have in poetry at present. These might resolve incorrectly. For example if provider searches for `(A git rev master)` and `(A git rev v1.0.0)` the cache would incorrectly return the package for the first one if I am groking the above correctly as the dict is using `hash(dep)`.

Reply:
Nope, `set` and `dict` are safe considering hash collisions. A set can hold different objects with same hashes.

See extreme example:

```
class A:
    def __init__(self, a):
        self.a = a

    def __hash__(self):
        return 0

    def __eq__(self, other):
        return self.a == other.a

    def __repr__(self):
        return f'A({self.a})'


s = {A(0), A(1), A(2)}
d = {A(0): 0, A(1): 1, A(2): 2}
print(s)
print(d[A(0)])
print(d[A(2)])
```

Output:
```
{A(0), A(1), A(2)}
0
2
```